integration/spark/vendor/iceberg/src/main/java/io/openlineage/spark/agent/vendor/iceberg/metrics/IcebergMetricsReporterInjector.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+
+/**
+ * Declared as a QueryPlanVisitor to be able to inject the InMemoryMetricsReporter into the Iceberg
+ * catalog. Registering metric reporter allows enriching OpenLineage events with ScanReport and
+ * CommitReport metrics.
+ *
+ * @param <D>
+ */
+@Slf4j
+public class IcebergMetricsReporterInjector<D extends OpenLineage.Dataset>
+    extends QueryPlanVisitor<LogicalPlan, D> {
+
+  private static final String ICEBERG_REPORTER_DISABLED = "iceberg.metricsReporterDisabled";
+
+  public IcebergMetricsReporterInjector(OpenLineageContext context) {
+    super(context);
+  }
+
+  @Override
+  public boolean isDefinedAt(LogicalPlan plan) {
+    // if this is called, then Iceberg classes are on the classpath
+
+    if (Optional.ofNullable(context.getOpenLineageConfig())
+        .map(SparkOpenLineageConfig::getVendors)
+        .map(VendorsConfig::getAdditionalProperties)
+        .map(p -> p.getOrDefault(ICEBERG_REPORTER_DISABLED, "false"))
+        .orElse("false")
+        .equalsIgnoreCase("true")) {
+      log.debug("Iceberg metrics reporter is disabled");
+      return false;
+    }
+
+    Optional<TableCatalog> catalog = getCatalog(plan);
+
+    // check if plan has a catalog field
+    if (!catalog.isPresent()) {
+      return false;
+    }
+
+    // check catalog class starts with org.apache.iceberg
+    return catalog.get().getClass().getCanonicalName().startsWith("org.apache.iceberg");
+  }
+
+  private Optional<TableCatalog> getCatalog(LogicalPlan plan) {
+    try {
+      if (MethodUtils.getAccessibleMethod(plan.getClass(), "name") == null) {
+        // try calling catalog method directly
+        return Optional.ofNullable(
+            (TableCatalog) MethodUtils.invokeMethod(plan, "catalog", (Object[]) null));
+      } else {
+        // call catalog method on name LogicalPlan
+        LogicalPlan namePlan =
+            (LogicalPlan) MethodUtils.invokeMethod(plan, "name", (Object[]) null);
+
+        if (MethodUtils.getAccessibleMethod(namePlan.getClass(), "catalog") == null) {
+          return Optional.empty();
+        }
+
+        return Optional.ofNullable(
+            (TableCatalog) MethodUtils.invokeMethod(namePlan, "catalog", (Object[]) null));
+      }
+    } catch (NoSuchMethodException e) {
+      // do nothing, don't log
+      return Optional.empty();
+    } catch (InvocationTargetException | IllegalAccessException e) {
+      log.debug("Could not find catalog in plan", e);
+      return Optional.empty();
+    }
+  }
+
+  @Override
+  public List<D> apply(LogicalPlan x) {
+    SparkCatalog catalog = (SparkCatalog) getCatalog(x).get();
+    // hack catalog to inject OpenLineageMetricsReporter
+    Catalog icebergCatalog = catalog.icebergCatalog();
+    if (icebergCatalog instanceof CachingCatalog) {
+      // get root catalog of a caching catalog
+      Field catalogField = FieldUtils.getField(icebergCatalog.getClass(), "catalog", true);
```

Comment:
Same here, this reflection behavior seems crucial to understand what's going on here, and is pretty opaque when all of those could fail, or why are we even using reflection instead of direct calls here.